crates/wasmparser/src/validator/component.rs

```diff
@@ -2170,10 +2306,11 @@ impl ComponentState {
                 CanonicalOption::Realloc(idx) => {
                     realloc = match realloc {
                         None => {
-                            let ty = types[self.core_function_at(*idx, offset)?].unwrap_func();
-                            if ty.params()
+                            let ty_id = self.core_function_at(*idx, offset)?;
+                            let func_ty = types[ty_id].unwrap_func();
+                            if func_ty.params()
                                 != [ValType::I32, ValType::I32, ValType::I32, ValType::I32]
-                                || ty.results() != [ValType::I32]
+                                || func_ty.results() != [ValType::I32]
                             {
```

Comment:
Do you think it might make sense to defer this type-check until `CanonicalOptions::check_{lower,lift}`? That's how post-return and callback options work for somewhat unrelated reasons, but in the future with memory64 integration it might be required for realloc as well

Reply:
I was trying to validate as much up front as possible, so that it is harder to forget to add things in the implementation of subsequent methods and so that it can be shared as much as possible. That said, I don't feel super strongly about it. My attitude is probably best summarized as

> maybe? but it isn't clear to me. let's wait and cross that bridge when we get to it